FocusController::focusedOrMainFrame should return a LocalFrame*

achristensen07 opened the pull request:
#### b99b4d6f85701aeb323b540fb5eaef0d891a564a
<pre>
FocusController::focusedOrMainFrame should return a LocalFrame*
<a href="https://bugs.webkit.org/show_bug.cgi?id=269761">https://bugs.webkit.org/show_bug.cgi?id=269761</a>
<a href="https://rdar.apple.com/116201648">rdar://116201648</a>

Reviewed by Chris Dumez.

With site isolation, the main frame is not always a LocalFrame.
Instead of returning a random root frame, correctly return null.
This requires null checks for all callers.
This should fix <a href="https://rdar.apple.com/121960445">rdar://121960445</a> too.

In order to make it so that the layout test security/clip-invalid-rect.html
does not crash because of an invalid IPC message WebPageProxy::EditorStateChanged,
I made it so that WebPage::editorState always returns an EditorState with a valid
identifier.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::focusedObjectForPage):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::focus):
(WebCore::Element::blur):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::selectFrameElementInParentIfFullySelected):
(WebCore::FrameSelection::setFocusedElementIfNeeded):
* Source/WebCore/editing/VisibleSelection.cpp:
(WebCore::VisibleSelection::emptySelection):
* Source/WebCore/editing/VisibleSelection.h:
* Source/WebCore/editing/cocoa/DictionaryLookup.mm:
(WebCore::DictionaryLookup::rangeAtHitTestResult):
* Source/WebCore/history/CachedPage.cpp:
(WebCore::CachedPage::restore):
* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::handleFocusEvent):
* Source/WebCore/inspector/InspectorFrontendHost.cpp:
(WebCore::InspectorFrontendHost::killText):
* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::performDragOperation):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handlePasteGlobalSelection):
(WebCore::EventHandler::dispatchMouseEvent):
(WebCore::EventHandler::internalKeyEvent):
(WebCore::EventHandler::defaultArrowEventHandler):
(WebCore::EventHandler::defaultTabEventHandler):
(WebCore::EventHandler::focusDocumentView):
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::focusedOrMainFrame const):
(WebCore::FocusController::setFocusedInternal):
(WebCore::FocusController::setInitialFocus):
(WebCore::FocusController::relinquishFocusToChrome):
(WebCore::FocusController::advanceFocusInDocumentOrder):
(WebCore::FocusController::setActiveInternal):
(WebCore::FocusController::advanceFocusDirectionallyInContainer):
(WebCore::FocusController::advanceFocusDirectionally):
(WebCore::FocusController::focusRepaintTimerFired):
* Source/WebCore/page/FocusController.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::focus):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::replaceSelectionWithText):
(WebCore::Page::editableElementsInRect const):
(WebCore::Page::selection const):
(WebCore::Page::fragmentDirectiveURLForSelectedText):
(WebCore::Page::revealCurrentSelection):
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::focusDocumentView):
* Source/WebCore/page/mac/EventHandlerMac.mm:
(WebCore::EventHandler::focusDocumentView):
* Source/WebCore/page/mac/ServicesOverlayController.mm:
(WebCore::ServicesOverlayController::remainingTimeUntilHighlightShouldBeShown const):
(WebCore::ServicesOverlayController::telephoneNumberRangesForFocusedFrame):
(WebCore::ServicesOverlayController::handleClick):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFrame.cpp:
(WKBundleFrameFocus):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKAccessibilityFocusedObject):
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebEditorClientMac.mm:
(WebKit::changeWordCase):
* Source/WebKit/WebProcess/WebPage/Cocoa/TextCheckingControllerProxy.mm:
(WebKit::TextCheckingControllerProxy::rangeAndOffsetRelativeToSelection):
(WebKit::TextCheckingControllerProxy::replaceRelativeToSelection):
(WebKit::TextCheckingControllerProxy::removeAnnotationRelativeToSelection):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::performDictionaryLookupAtLocation):
(WebKit::WebPage::performDictionaryLookupOfCurrentSelection):
(WebKit::WebPage::insertDictatedTextAsync):
(WebKit::WebPage::addDictationAlternative):
(WebKit::WebPage::dictationAlternativesAtSelection):
(WebKit::WebPage::clearDictationAlternatives):
(WebKit::WebPage::replaceImageForRemoveBackground):
(WebKit::WebPage::readSelectionFromPasteboard):
(WebKit::WebPage::insertMultiRepresentationHEIC):
(WebKit::WebPage::autocorrectionContextRange):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
(WebKit::UnifiedTextReplacementController::willBeginTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveReplacements):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement):
(WebKit::UnifiedTextReplacementController::didEndTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveTextWithReplacementRange):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveEditAction):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::createTextIndicatorForRange):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::editorState const):
(WebKit::WebPage::changeFontAttributes):
(WebKit::WebPage::changeFont):
(WebKit::WebPage::updateEditorStateAfterLayoutIfEditabilityChanged):
(WebKit::WebPage::executeEditingCommand):
(WebKit::WebPage::setEditable):
(WebKit::WebPage::increaseListLevel):
(WebKit::WebPage::decreaseListLevel):
(WebKit::WebPage::changeListType):
(WebKit::WebPage::setBaseWritingDirection):
(WebKit::WebPage::isEditingCommandEnabled):
(WebKit::WebPage::contextMenuForKeyEvent):
(WebKit::WebPage::handleKeyEventByRelinquishingFocusToChrome):
(WebKit::WebPage::validateCommand):
(WebKit::WebPage::requestFontAttributesAtSelectionStart):
(WebKit::WebPage::scroll):
(WebKit::WebPage::logicalScroll):
(WebKit::WebPage::centerSelectionInVisibleArea):
(WebKit::WebPage::insertNewlineInQuotedContent):
(WebKit::WebPage::viewWillStartLiveResize):
(WebKit::WebPage::viewWillEndLiveResize):
(WebKit::WebPage::setInitialFocus):
(WebKit::WebPage::copyLinkToHighlight):
(WebKit::WebPage::getSelectionOrContentsAsString):
(WebKit::WebPage::advanceToNextMisspelling):
(WebKit::WebPage::hasRichlyEditableSelection const):
(WebKit::WebPage::changeSpellingToWord):
(WebKit::WebPage::uppercaseWord):
(WebKit::WebPage::lowercaseWord):
(WebKit::WebPage::capitalizeWord):
(WebKit::WebPage::clearSelection):
(WebKit::WebPage::restoreSelectionInFocusedEditableElement):
(WebKit::WebPage::setCaretAnimatorType):
(WebKit::WebPage::setCaretBlinkingSuspended):
(WebKit::WebPage::handleAlternativeTextUIResult):
(WebKit::WebPage::setCompositionForTesting):
(WebKit::WebPage::hasCompositionForTesting):
(WebKit::WebPage::confirmCompositionForTesting):
(WebKit::WebPage::setTextAsync):
(WebKit::WebPage::insertTextAsync):
(WebKit::WebPage::hasMarkedText):
(WebKit::WebPage::getMarkedRangeAsync):
(WebKit::WebPage::getSelectedRangeAsync):
(WebKit::WebPage::characterIndexForPointAsync):
(WebKit::WebPage::firstRectForCharacterRangeAsync):
(WebKit::WebPage::setCompositionAsync):
(WebKit::WebPage::confirmCompositionAsync):
(WebKit::targetFrameForEditing):
(WebKit::WebPage::didChangeSelectionOrOverflowScrollPosition):
(WebKit::WebPage::didEndUserTriggeredSelectionChanges):
(WebKit::WebPage::sendEditorStateUpdate):
(WebKit::WebPage::flushPendingEditorStateUpdate):
(WebKit::WebPage::insertAttachment):
(WebKit::WebPage::createAppHighlightInSelectedRange):
(WebKit::WebPage::restoreAppHighlightsAndScrollToIndex):
* Source/WebKit/WebProcess/WebPage/ios/FindControllerIOS.mm:
(WebKit::FindController::didFindString):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::platformReinitialize):
(WebKit::WebPage::platformWillPerformEditingCommand):
(WebKit::WebPage::getSelectionContext):
(WebKit::WebPage::updateSelectionAppearance):
(WebKit::WebPage::generateSyntheticEditingCommand):
(WebKit::WebPage::completeSyntheticClick):
(WebKit::WebPage::didConcludeEditDrag):
(WebKit::WebPage::clearSelectionAfterTapIfNeeded):
(WebKit::WebPage::selectWithGesture):
(WebKit::WebPage::clearSelection):
(WebKit::WebPage::dispatchSyntheticMouseEventsForSelectionGesture):
(WebKit::WebPage::updateSelectionWithTouches):
(WebKit::WebPage::selectWithTwoTouches):
(WebKit::WebPage::extendSelectionForReplacement):
(WebKit::WebPage::extendSelection):
(WebKit::WebPage::setSelectedRangeDispatchingSyntheticMouseEventsIfNeeded):
(WebKit::WebPage::platformDidSelectAll):
(WebKit::WebPage::selectWordBackward):
(WebKit::WebPage::moveSelectionByOffset):
(WebKit::WebPage::startAutoscrollAtPosition):
(WebKit::WebPage::requestEvasionRectsAboveSelection):
(WebKit::WebPage::getRectsForGranularityWithSelectionOffset):
(WebKit::WebPage::storeSelectionForAccessibility):
(WebKit::WebPage::getRectsAtSelectionOffsetWithText):
(WebKit::WebPage::selectPositionAtPoint):
(WebKit::WebPage::selectPositionAtBoundaryWithDirection):
(WebKit::WebPage::moveSelectionAtBoundaryWithDirection):
(WebKit::WebPage::setFocusedFrameBeforeSelectingTextAtLocation):
(WebKit::WebPage::setSelectionRange):
(WebKit::WebPage::updateSelectionWithExtentPointAndBoundary):
(WebKit::WebPage::updateSelectionWithExtentPoint):
(WebKit::WebPage::revealItemForCurrentSelection):
(WebKit::WebPage::replaceSelectedText):
(WebKit::WebPage::replaceDictatedText):
(WebKit::WebPage::willInsertFinalDictationResult):
(WebKit::WebPage::requestAutocorrectionData):
(WebKit::WebPage::applyAutocorrectionInternal):
(WebKit::WebPage::autocorrectionContext):
(WebKit::focusedElementPositionInformation):
(WebKit::WebPage::focusedElementInformation):
(WebKit::WebPage::hardwareKeyboardAvailabilityChanged):
(WebKit::WebPage::insertTextPlaceholder):
(WebKit::WebPage::updateSelectionWithDelta):
(WebKit::WebPage::requestDocumentEditingContext):
(WebKit::WebPage::focusTextInputContextAndPlaceCaret):
(WebKit::WebPage::animationDidFinishForElement):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::platformReinitialize):
(WebKit::WebPage::executeKeypressCommandsInternal):
(WebKit::WebPage::attributedSubstringForCharacterRangeAsync):
(WebKit::WebPage::getStringSelectionForPasteboard):
(WebKit::WebPage::getDataSelectionForPasteboard):
(WebKit::WebPage::shouldDelayWindowOrderingEvent):
(WebKit::WebPage::requestAcceptsFirstMouse):
(WebKit::WebPage::performImmediateActionHitTestAtLocation):
* Source/WebKitLegacy/mac/Misc/WebSharingServicePickerController.mm:
(-[WebSharingServicePickerController sharingService:didShareItems:]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _didConcludeEditDrag]):
(-[WebView _executeCoreCommandByName:value:]):

Canonical link: <a href="https://commits.webkit.org/275171@main">https://commits.webkit.org/275171@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e86cdd5f1f3e9d718a33bfdb6f5effc9b084d76a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41050 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20063 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43428 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43610 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37142 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43357 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23080 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17394 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33990 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41624 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16985 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35359 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14622 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14737 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36357 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44922 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37240 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36669 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40418 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15865 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13009 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38795 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17484 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9223 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17535 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17128 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -481,16 +481,13 @@ void FocusController::setFocusedFrame(Frame* frame, BroadcastFocusedFrame broadc
 }
 
-LocalFrame& FocusController::focusedOrMainFrame() const
+RefPtr<LocalFrame> FocusController::focusedOrMainFrame() const
```
This goes against our usual patterns. I think this should return a `LocalFrame*`, not a RefPtr. This doesn't transfer ownership and we don't want to cause unnecessary churn.

cdumez replied:
cc @rniwa

cdumez replied:
getters that return a RefPtr are named `protectedFoo()` and are only used to avoid local variables:
e.g.
```
focusController->protectedFocusedOrMainFrame()->doNonTrivialStuff();
```

We'd still use the `focusedOrMainFrame()` one that returns a raw pointer in those cases:
```
RefPtr frame = focusController->focusedOrMainFrame();
```
or
```
return focusController->focusedOrMainFrame()->trivialInlineGetter(); // To avoid unnecessary refcounting churn.
```

achristensen07 replied:
Ok.  The last one would definitely need a null check.  That's the whole point of this PR.  Sometimes with site isolation we neither have a local main frame nor a focused frame.

cdumez replied:
Yes, but is doesn't matter, the pointer can stay raw in that case.

cdumez commented on Source/WebCore/editing/VisibleSelection.cpp:
```diff
@@ -37,13 +37,20 @@
 
 namespace WebCore {
 
+VisibleSelection& VisibleSelection::emptySelection()
```
Presumably, this should return a `const VisibleSelection&`

cdumez commented on Source/WebCore/editing/cocoa/DictionaryLookup.mm:
```diff
@@ -313,7 +313,11 @@ static bool canCreateRevealItems()
         position = firstPositionInOrBeforeNode(node);
 
-    auto selection = CheckedRef(frame->page()->focusController())->focusedOrMainFrame().selection().selection();
+    RefPtr focusedOrMainFrame = CheckedRef(frame->page()->focusController())->focusedOrMainFrame();
```
`frame->page()->checkedFocusController()->focusedOrMainFrame()` is what it's supposed to be. This code is a bit outdated and it'd be nice to refresh it.

cdumez commented on Source/WebCore/history/CachedPage.cpp:
```diff
@@ -134,7 +134,11 @@ void CachedPage::restore(Page& page)
     // Restore the focus appearance for the focused element.
     // FIXME: Right now we don't support pages w/ frames in the b/f cache.  This may need to be tweaked when we add support for that.
-    RefPtr focusedDocument = CheckedRef(page.focusController())->focusedOrMainFrame().document();
+    RefPtr focusedOrMainFrame = CheckedRef(page.focusController())->focusedOrMainFrame();
```
Same comment about calling checkedFocusController()

cdumez commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2403,9 +2403,9 @@ bool EventHandler::handlePasteGlobalSelection(const PlatformMouseEvent& platform
     if (!m_frame->page())
         return false;
-    LocalFrame& focusFrame = CheckedRef(m_frame->page()->focusController())->focusedOrMainFrame();
+    RefPtr focusFrame = CheckedRef(m_frame->page()->focusController())->focusedOrMainFrame();
```
Same comment about calling checkedFocusController()

cdumez commented on Source/WebCore/inspector/InspectorFrontendHost.cpp:
```diff
@@ -442,7 +442,11 @@ void InspectorFrontendHost::killText(const String& text, bool shouldPrependToKil
         return;
 
-    Editor& editor = m_frontendPage->focusController().focusedOrMainFrame().editor();
+    RefPtr focusedOrMainFrame = m_frontendPage->focusController().focusedOrMainFrame();
```
It'd be nice to call checkedFocusController() for consistency.

cdumez commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -238,7 +238,11 @@ bool DragController::performDragOperation(DragData&& dragData)
 
     SetForScope isPerformingDrop(m_isPerformingDrop, true);
-    IgnoreSelectionChangeForScope ignoreSelectionChanges { m_page->focusController().focusedOrMainFrame() };
+    RefPtr focusedOrMainFrame = m_page->focusController().focusedOrMainFrame();
```
`checkedFocusController()` for consistency.

cdumez commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -481,16 +481,13 @@ void FocusController::setFocusedFrame(Frame* frame, BroadcastFocusedFrame broadc
-        return *localMainFrame;
-
-    // FIXME: Do something better here in the site isolated case. <rdar://116201648>
-    ASSERT(m_page->settings().siteIsolationEnabled());
```
I think it would be nice to keep this assertion.

cdumez commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -500,8 +497,9 @@ void FocusController::setFocused(bool focused)
 {
-    if (!isFocused())
-        focusedOrMainFrame().eventHandler().stopAutoscrollTimer();
+    RefPtr focusedOrMainFrame = this->focusedOrMainFrame();
```
Can we put this inline the `if (!isFocused())` scope to avoid unnecessary ref counting churn when we're already focused?

cdumez commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -1222,14 +1230,18 @@ bool FocusController::advanceFocusDirectionallyInContainer(Node* container, cons
 
     // Figure out the starting rect.
-    Node* container = focusedDocument;
+    RefPtr<Node> container = focusedDocument.get();
```
I believe `.get()` is unnecessary

achristensen07 replied:
While you're correct, the style bot complains if it's removed.  Since it's equivalent, I'm going to leave it.

cdumez replied:
Weird decision..

cdumez replied:
Let's make write more unnecessary code because of a style bot false positive?
I am also curious what the style bot finds wrong when omitting `.get()`.

cdumez commented on Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
```diff
@@ -287,7 +287,10 @@ void* WKAccessibilityFocusedObject(WKBundlePageRef pageRef)
-    RefPtr focusedDocument = CheckedRef(page->focusController())->focusedOrMainFrame().document();
+    RefPtr focusedOrMainFrame = CheckedRef(page->focusController())->focusedOrMainFrame();
+    if (!focusedOrMainFrame)
+        return 0;
```
Since you're touching this function, some `nullptr` conversion would have been nice.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/mac/WebEditorClientMac.mm:
```diff
@@ -67,14 +67,16 @@
 static void changeWordCase(WebPage* page, NSString *(*changeCase)(NSString *))
 {
-    Ref frame = CheckedRef(page->corePage()->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(page->corePage()->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`, no need to `CheckedRef()`

cdumez commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextCheckingControllerProxy.mm:
```diff
@@ -158,7 +161,10 @@ typedef NS_ENUM(NSInteger, NSSpellingState) {
     auto removeCoreSpellingMarkers = annotation == "NSSpellingState"_s;
     auto types = removeCoreSpellingMarkers ? relevantMarkerTypes() : WebCore::DocumentMarker::Type::PlatformTextChecking;
-    RefPtr document = CheckedRef(m_page.corePage()->focusController())->focusedOrMainFrame().document();
+    RefPtr focusedOrMainFrame = CheckedRef(m_page.corePage()->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -202,7 +202,7 @@
     auto result = localMainFrame->eventHandler().hitTestResultAtPoint(localMainFrame->view()->windowToContents(roundedIntPoint(floatPoint)), hitType);
 
-    RefPtr frame = result.innerNonSharedNode() ? result.innerNonSharedNode()->document().frame() : &CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = result.innerNonSharedNode() ? result.innerNonSharedNode()->document().frame() : CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -224,8 +224,11 @@
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
-    performDictionaryLookupForSelection(frame, frame->selection().selection(), TextIndicatorPresentationTransition::BounceAndCrossfade);
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -298,10 +301,12 @@
 void WebPage::insertDictatedTextAsync(const String& text, const EditingRange& replacementEditingRange, const Vector<WebCore::DictationAlternative>& dictationAlternativeLocations, InsertTextOptions&& options)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -326,7 +331,10 @@
 void WebPage::addDictationAlternative(const String& text, DictationContext context, CompletionHandler<void(bool)>&& completion)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -362,7 +370,10 @@
 void WebPage::dictationAlternativesAtSelection(CompletionHandler<void(Vector<DictationContext>&&)>&& completion)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -387,7 +398,10 @@
 void WebPage::clearDictationAlternatives(Vector<DictationContext>&& contexts)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -693,7 +707,10 @@
 void WebPage::replaceImageForRemoveBackground(const ElementContext& elementContext, const Vector<String>& types, std::span<const uint8_t> data)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1434,15 +1434,17 @@ bool WebPage::hasPendingEditorStateUpdate() const
 {
     // Ref the frame because this function may perform layout, which may cause frame destruction.
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1497,21 +1499,27 @@ EditorState WebPage::editorState(ShouldPerformLayout shouldPerformLayout) const
 void WebPage::changeFontAttributes(WebCore::FontAttributeChanges&& changes)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1497,21 +1499,27 @@ EditorState WebPage::editorState(ShouldPerformLayout shouldPerformLayout) const
 void WebPage::changeFont(WebCore::FontChanges&& changes)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1563,7 +1571,10 @@ void WebPage::updateEditorStateAfterLayoutIfEditabilityChanged()
         return;
 
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1686,10 +1697,12 @@ void WebPage::executeEditingCommand(const String& commandName, const String& arg
     platformWillPerformEditingCommand();
 
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1702,7 +1715,10 @@ void WebPage::setEditable(bool editable)
     m_page->setEditable(editable);
     m_page->setTabKeyCyclesThroughElements(!editable);
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1713,34 +1729,48 @@ void WebPage::setEditable(bool editable)
 void WebPage::increaseListLevel()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1713,34 +1729,48 @@ void WebPage::setEditable(bool editable)
 void WebPage::decreaseListLevel()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1713,34 +1729,48 @@ void WebPage::setEditable(bool editable)
 void WebPage::changeListType()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1713,34 +1729,48 @@ void WebPage::setEditable(bool editable)
 void WebPage::setBaseWritingDirection(WritingDirection direction)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1713,34 +1729,48 @@ void WebPage::setEditable(bool editable)
 bool WebPage::isEditingCommandEnabled(const String& commandName)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3410,7 +3440,10 @@ void WebPage::contextMenuForKeyEvent()
 #endif
 
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3617,10 +3650,12 @@ void WebPage::validateCommand(const String& commandName, CompletionHandler<void(
     bool isEnabled = false;
     int32_t state = 0;
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3682,7 +3717,10 @@ void WebPage::setCurrentHistoryItemForReattach(WebKit::BackForwardListItemState&
 void WebPage::requestFontAttributesAtSelectionStart(CompletionHandler<void(const WebCore::FontAttributes&)>&& completionHandler)
 {
-    completionHandler(CheckedRef(m_page->focusController())->focusedOrMainFrame().editor().fontAttributesAtSelectionStart());
+    RefPtr focusedOrMainFrame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3819,7 +3863,9 @@ bool WebPage::scrollBy(WebCore::ScrollDirection scrollDirection, WebCore::Scroll
 void WebPage::centerSelectionInVisibleArea()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3851,7 +3897,9 @@ void WebPage::sendMessageToTargetBackend(const String& targetId, const String& m
 void WebPage::insertNewlineInQuotedContent()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3904,7 +3952,10 @@ void WebPage::viewWillStartLiveResize()
 
     // FIXME: This should propagate to all ScrollableAreas.
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3915,7 +3966,10 @@ void WebPage::viewWillEndLiveResize()
 
     // FIXME: This should propagate to all ScrollableAreas.
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -4339,14 +4395,18 @@ void WebPage::getSelectionAsWebArchiveData(CompletionHandler<void(const std::opt
 {
     auto url = m_page->fragmentDirectiveURLForSelectedText();
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -4339,14 +4395,18 @@ void WebPage::getSelectionAsWebArchiveData(CompletionHandler<void(const std::opt
 void WebPage::getSelectionOrContentsAsString(CompletionHandler<void(const String&)>&& callback)
 {
-    RefPtr focusedOrMainFrame = WebFrame::fromCoreFrame(CheckedRef(m_page->focusController())->focusedOrMainFrame());
+    RefPtr focusedOrMainCoreFrame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5658,14 +5718,20 @@ void WebPage::mediaKeySystemWasDenied(MediaKeySystemRequestIdentifier mediaKeySy
 void WebPage::advanceToNextMisspelling(bool startBeforeSelection)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5658,14 +5718,20 @@ void WebPage::mediaKeySystemWasDenied(MediaKeySystemRequestIdentifier mediaKeySy
 bool WebPage::hasRichlyEditableSelection() const
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5674,7 +5740,7 @@ bool WebPage::hasRichlyEditableSelection() const
 void WebPage::changeSpellingToWord(const String& word)
 {
-    replaceSelectionWithText(&CheckedRef(m_page->focusController())->focusedOrMainFrame(), word);
+    replaceSelectionWithText(CheckedRef(m_page->focusController())->focusedOrMainFrame(), word);
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5702,19 +5768,28 @@ void WebPage::unmarkAllBadGrammar()
 void WebPage::uppercaseWord()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5702,19 +5768,28 @@ void WebPage::unmarkAllBadGrammar()
 void WebPage::lowercaseWord()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5702,19 +5768,28 @@ void WebPage::unmarkAllBadGrammar()
 void WebPage::capitalizeWord()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5753,14 +5828,20 @@ void WebPage::replaceSelectionWithText(LocalFrame* frame, const String& text)
 void WebPage::clearSelection()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5753,14 +5828,20 @@ void WebPage::replaceSelectionWithText(LocalFrame* frame, const String& text)
 void WebPage::restoreSelectionInFocusedEditableElement()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6082,13 +6163,19 @@ void WebPage::stopSpeaking()
 void WebPage::setCaretAnimatorType(WebCore::CaretAnimatorType caretType)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6082,13 +6163,19 @@ void WebPage::stopSpeaking()
 void WebPage::setCaretBlinkingSuspended(bool suspended)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6486,14 +6573,20 @@ bool WebPage::canHandleRequest(const WebCore::ResourceRequest& request)
 void WebPage::handleAlternativeTextUIResult(const String& result)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6486,14 +6573,20 @@ bool WebPage::canHandleRequest(const WebCore::ResourceRequest& request)
 void WebPage::setCompositionForTesting(const String& compositionString, uint64_t from, uint64_t length, bool suppressUnderline, const Vector<CompositionHighlight>& highlights, const HashMap<String, Vector<WebCore::CharacterRange>>& annotations)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6506,13 +6599,19 @@ void WebPage::setCompositionForTesting(const String& compositionString, uint64_t
 bool WebPage::hasCompositionForTesting()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6506,13 +6599,19 @@ void WebPage::setCompositionForTesting(const String& compositionString, uint64_t
 void WebPage::confirmCompositionForTesting(const String& compositionString)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6629,9 +6728,12 @@ bool WebPage::shouldUseCustomContentProviderForResponse(const ResourceResponse&
 void WebPage::setTextAsync(const String& text)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6652,13 +6754,15 @@ void WebPage::insertTextAsync(const String& text, const EditingRange& replacemen
     platformWillPerformEditingCommand();
 
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6691,19 +6795,28 @@ void WebPage::insertTextAsync(const String& text, const EditingRange& replacemen
 void WebPage::hasMarkedText(CompletionHandler<void(bool)>&& completionHandler)
 {
-    completionHandler(CheckedRef(m_page->focusController())->focusedOrMainFrame().editor().hasComposition());
+    RefPtr focusedOrMainFrame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6691,19 +6795,28 @@ void WebPage::insertTextAsync(const String& text, const EditingRange& replacemen
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
-    completionHandler(EditingRange::fromRange(frame, frame->editor().compositionRange()));
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6691,19 +6795,28 @@ void WebPage::insertTextAsync(const String& text, const EditingRange& replacemen
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
-    completionHandler(EditingRange::fromRange(frame, frame->selection().selection().toNormalizedRange()));
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6713,16 +6826,21 @@ void WebPage::characterIndexForPointAsync(const WebCore::IntPoint& point, Comple
     constexpr OptionSet<HitTestRequest::Type> hitType { HitTestRequest::Type::ReadOnly, HitTestRequest::Type::Active, HitTestRequest::Type::DisallowUserAgentShadowContent,  HitTestRequest::Type::AllowChildFrameContent };
     auto result = localMainFrame->eventHandler().hitTestResultAtPoint(point, hitType);
-    Ref frame = result.innerNonSharedNode() ? *result.innerNodeFrame() : CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = result.innerNonSharedNode() ? result.innerNodeFrame() : CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6713,16 +6826,21 @@ void WebPage::characterIndexForPointAsync(const WebCore::IntPoint& point, Comple
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
-    auto range = EditingRange::toRange(frame, editingRange);
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6735,10 +6853,13 @@ void WebPage::setCompositionAsync(const String& text, const Vector<CompositionUn
     platformWillPerformEditingCommand();
 
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6749,7 +6870,10 @@ void WebPage::confirmCompositionAsync()
     platformWillPerformEditingCommand();
 
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6848,7 +6972,10 @@ void WebPage::didChangeSelection(LocalFrame& frame)
 void WebPage::didChangeSelectionOrOverflowScrollPosition()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7041,7 +7168,10 @@ void WebPage::didUpdateComposition()
 void WebPage::didEndUserTriggeredSelectionChanges()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7540,7 +7673,10 @@ void WebPage::flushPendingEditorStateUpdate()
         return;
 
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8103,7 +8239,10 @@ void WebPage::voicesDidChange()
 void WebPage::insertAttachment(const String& identifier, std::optional<uint64_t>&& fileSize, const String& fileName, const String& contentType, CompletionHandler<void()>&& callback)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8781,7 +8920,10 @@ bool WebPage::createAppHighlightInSelectedRange(WebCore::CreateNewGroupForHighli
     SetForScope highlightRequestOriginScope { m_highlightRequestOriginatedInApp, requestOriginatedInApp };
 
-    RefPtr document = CheckedRef(m_page->focusController())->focusedOrMainFrame().document();
+    RefPtr focusedOrMainFrame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8799,7 +8941,10 @@ bool WebPage::createAppHighlightInSelectedRange(WebCore::CreateNewGroupForHighli
 void WebPage::restoreAppHighlightsAndScrollToIndex(Vector<SharedMemory::Handle>&& memoryHandles, const std::optional<unsigned> index)
 {
-    RefPtr document = CheckedRef(m_page->focusController())->focusedOrMainFrame().document();
+    RefPtr focusedOrMainFrame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -250,7 +250,10 @@ bool scalesAreEssentiallyEqual(float a, float b)
 void WebPage::platformReinitialize()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -423,7 +426,10 @@ static void convertContentToRootView(const LocalFrameView& view, Vector<Selectio
 {
 #if ENABLE(CONTENT_CHANGE_OBSERVER)
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -629,7 +635,10 @@ static inline FloatRect adjustExposedRectForNewScale(const FloatRect& exposedRec
 void WebPage::getSelectionContext(CompletionHandler<void(const String&, const String&, const String&)>&& completionHandler)
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -728,7 +737,10 @@ static inline FloatRect adjustExposedRectForNewScale(const FloatRect& exposedRec
 void WebPage::updateSelectionAppearance()
 {
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -753,7 +765,9 @@ static void dispatchSyntheticMouseMove(LocalFrame& mainFrame, const WebCore::Flo
 {
     PlatformKeyboardEvent keyEvent;
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -1077,7 +1091,10 @@ static void dispatchSyntheticMouseMove(LocalFrame& mainFrame, const WebCore::Flo
     m_pendingImageElementsForDropSnapshot.clear();
 
-    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    RefPtr frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```
`checkedFocusController()`

cdumez replied:
Ok, I'm going to stop commenting on these, there are too many.

achristensen07 committed:
FocusController::focusedOrMainFrame should return a LocalFrame*
https://bugs.webkit.org/show_bug.cgi?id=269761
rdar://116201648

Reviewed by Chris Dumez.

With site isolation, the main frame is not always a LocalFrame.
Instead of returning a random root frame, correctly return null.
This requires null checks for all callers.
This should fix rdar://121960445 too.

In order to make it so that the layout test security/clip-invalid-rect.html
does not crash because of an invalid IPC message WebPageProxy::EditorStateChanged,
I made it so that WebPage::editorState always returns an EditorState with a valid
identifier.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::focusedObjectForPage):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::focus):
(WebCore::Element::blur):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::selectFrameElementInParentIfFullySelected):
(WebCore::FrameSelection::setFocusedElementIfNeeded):
* Source/WebCore/editing/VisibleSelection.cpp:
(WebCore::VisibleSelection::emptySelection):
* Source/WebCore/editing/VisibleSelection.h:
* Source/WebCore/editing/cocoa/DictionaryLookup.mm:
(WebCore::DictionaryLookup::rangeAtHitTestResult):
* Source/WebCore/history/CachedPage.cpp:
(WebCore::CachedPage::restore):
* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::handleFocusEvent):
* Source/WebCore/inspector/InspectorFrontendHost.cpp:
(WebCore::InspectorFrontendHost::killText):
* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::performDragOperation):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handlePasteGlobalSelection):
(WebCore::EventHandler::dispatchMouseEvent):
(WebCore::EventHandler::internalKeyEvent):
(WebCore::EventHandler::defaultArrowEventHandler):
(WebCore::EventHandler::defaultTabEventHandler):
(WebCore::EventHandler::focusDocumentView):
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::focusedOrMainFrame const):
(WebCore::FocusController::setFocusedInternal):
(WebCore::FocusController::setInitialFocus):
(WebCore::FocusController::relinquishFocusToChrome):
(WebCore::FocusController::advanceFocusInDocumentOrder):
(WebCore::FocusController::setActiveInternal):
(WebCore::FocusController::advanceFocusDirectionallyInContainer):
(WebCore::FocusController::advanceFocusDirectionally):
(WebCore::FocusController::focusRepaintTimerFired):
* Source/WebCore/page/FocusController.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::focus):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::replaceSelectionWithText):
(WebCore::Page::editableElementsInRect const):
(WebCore::Page::selection const):
(WebCore::Page::fragmentDirectiveURLForSelectedText):
(WebCore::Page::revealCurrentSelection):
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::focusDocumentView):
* Source/WebCore/page/mac/EventHandlerMac.mm:
(WebCore::EventHandler::focusDocumentView):
* Source/WebCore/page/mac/ServicesOverlayController.mm:
(WebCore::ServicesOverlayController::remainingTimeUntilHighlightShouldBeShown const):
(WebCore::ServicesOverlayController::telephoneNumberRangesForFocusedFrame):
(WebCore::ServicesOverlayController::handleClick):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFrame.cpp:
(WKBundleFrameFocus):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKAccessibilityFocusedObject):
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebEditorClientMac.mm:
(WebKit::changeWordCase):
* Source/WebKit/WebProcess/WebPage/Cocoa/TextCheckingControllerProxy.mm:
(WebKit::TextCheckingControllerProxy::rangeAndOffsetRelativeToSelection):
(WebKit::TextCheckingControllerProxy::replaceRelativeToSelection):
(WebKit::TextCheckingControllerProxy::removeAnnotationRelativeToSelection):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::performDictionaryLookupAtLocation):
(WebKit::WebPage::performDictionaryLookupOfCurrentSelection):
(WebKit::WebPage::insertDictatedTextAsync):
(WebKit::WebPage::addDictationAlternative):
(WebKit::WebPage::dictationAlternativesAtSelection):
(WebKit::WebPage::clearDictationAlternatives):
(WebKit::WebPage::replaceImageForRemoveBackground):
(WebKit::WebPage::readSelectionFromPasteboard):
(WebKit::WebPage::insertMultiRepresentationHEIC):
(WebKit::WebPage::autocorrectionContextRange):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
(WebKit::UnifiedTextReplacementController::willBeginTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveReplacements):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement):
(WebKit::UnifiedTextReplacementController::didEndTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveTextWithReplacementRange):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveEditAction):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::createTextIndicatorForRange):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::editorState const):
(WebKit::WebPage::changeFontAttributes):
(WebKit::WebPage::changeFont):
(WebKit::WebPage::updateEditorStateAfterLayoutIfEditabilityChanged):
(WebKit::WebPage::executeEditingCommand):
(WebKit::WebPage::setEditable):
(WebKit::WebPage::increaseListLevel):
(WebKit::WebPage::decreaseListLevel):
(WebKit::WebPage::changeListType):
(WebKit::WebPage::setBaseWritingDirection):
(WebKit::WebPage::isEditingCommandEnabled):
(WebKit::WebPage::contextMenuForKeyEvent):
(WebKit::WebPage::handleKeyEventByRelinquishingFocusToChrome):
(WebKit::WebPage::validateCommand):
(WebKit::WebPage::requestFontAttributesAtSelectionStart):
(WebKit::WebPage::scroll):
(WebKit::WebPage::logicalScroll):
(WebKit::WebPage::centerSelectionInVisibleArea):
(WebKit::WebPage::insertNewlineInQuotedContent):
(WebKit::WebPage::viewWillStartLiveResize):
(WebKit::WebPage::viewWillEndLiveResize):
(WebKit::WebPage::setInitialFocus):
(WebKit::WebPage::copyLinkToHighlight):
(WebKit::WebPage::getSelectionOrContentsAsString):
(WebKit::WebPage::advanceToNextMisspelling):
(WebKit::WebPage::hasRichlyEditableSelection const):
(WebKit::WebPage::changeSpellingToWord):
(WebKit::WebPage::uppercaseWord):
(WebKit::WebPage::lowercaseWord):
(WebKit::WebPage::capitalizeWord):
(WebKit::WebPage::clearSelection):
(WebKit::WebPage::restoreSelectionInFocusedEditableElement):
(WebKit::WebPage::setCaretAnimatorType):
(WebKit::WebPage::setCaretBlinkingSuspended):
(WebKit::WebPage::handleAlternativeTextUIResult):
(WebKit::WebPage::setCompositionForTesting):
(WebKit::WebPage::hasCompositionForTesting):
(WebKit::WebPage::confirmCompositionForTesting):
(WebKit::WebPage::setTextAsync):
(WebKit::WebPage::insertTextAsync):
(WebKit::WebPage::hasMarkedText):
(WebKit::WebPage::getMarkedRangeAsync):
(WebKit::WebPage::getSelectedRangeAsync):
(WebKit::WebPage::characterIndexForPointAsync):
(WebKit::WebPage::firstRectForCharacterRangeAsync):
(WebKit::WebPage::setCompositionAsync):
(WebKit::WebPage::confirmCompositionAsync):
(WebKit::targetFrameForEditing):
(WebKit::WebPage::didChangeSelectionOrOverflowScrollPosition):
(WebKit::WebPage::didEndUserTriggeredSelectionChanges):
(WebKit::WebPage::sendEditorStateUpdate):
(WebKit::WebPage::flushPendingEditorStateUpdate):
(WebKit::WebPage::insertAttachment):
(WebKit::WebPage::createAppHighlightInSelectedRange):
(WebKit::WebPage::restoreAppHighlightsAndScrollToIndex):
* Source/WebKit/WebProcess/WebPage/ios/FindControllerIOS.mm:
(WebKit::FindController::didFindString):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::platformReinitialize):
(WebKit::WebPage::platformWillPerformEditingCommand):
(WebKit::WebPage::getSelectionContext):
(WebKit::WebPage::updateSelectionAppearance):
(WebKit::WebPage::generateSyntheticEditingCommand):
(WebKit::WebPage::completeSyntheticClick):
(WebKit::WebPage::didConcludeEditDrag):
(WebKit::WebPage::clearSelectionAfterTapIfNeeded):
(WebKit::WebPage::selectWithGesture):
(WebKit::WebPage::clearSelection):
(WebKit::WebPage::dispatchSyntheticMouseEventsForSelectionGesture):
(WebKit::WebPage::updateSelectionWithTouches):
(WebKit::WebPage::selectWithTwoTouches):
(WebKit::WebPage::extendSelectionForReplacement):
(WebKit::WebPage::extendSelection):
(WebKit::WebPage::setSelectedRangeDispatchingSyntheticMouseEventsIfNeeded):
(WebKit::WebPage::platformDidSelectAll):
(WebKit::WebPage::selectWordBackward):
(WebKit::WebPage::moveSelectionByOffset):
(WebKit::WebPage::startAutoscrollAtPosition):
(WebKit::WebPage::requestEvasionRectsAboveSelection):
(WebKit::WebPage::getRectsForGranularityWithSelectionOffset):
(WebKit::WebPage::storeSelectionForAccessibility):
(WebKit::WebPage::getRectsAtSelectionOffsetWithText):
(WebKit::WebPage::selectPositionAtPoint):
(WebKit::WebPage::selectPositionAtBoundaryWithDirection):
(WebKit::WebPage::moveSelectionAtBoundaryWithDirection):
(WebKit::WebPage::setFocusedFrameBeforeSelectingTextAtLocation):
(WebKit::WebPage::setSelectionRange):
(WebKit::WebPage::updateSelectionWithExtentPointAndBoundary):
(WebKit::WebPage::updateSelectionWithExtentPoint):
(WebKit::WebPage::revealItemForCurrentSelection):
(WebKit::WebPage::replaceSelectedText):
(WebKit::WebPage::replaceDictatedText):
(WebKit::WebPage::willInsertFinalDictationResult):
(WebKit::WebPage::requestAutocorrectionData):
(WebKit::WebPage::applyAutocorrectionInternal):
(WebKit::WebPage::autocorrectionContext):
(WebKit::focusedElementPositionInformation):
(WebKit::WebPage::focusedElementInformation):
(WebKit::WebPage::hardwareKeyboardAvailabilityChanged):
(WebKit::WebPage::insertTextPlaceholder):
(WebKit::WebPage::updateSelectionWithDelta):
(WebKit::WebPage::requestDocumentEditingContext):
(WebKit::WebPage::focusTextInputContextAndPlaceCaret):
(WebKit::WebPage::animationDidFinishForElement):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::platformReinitialize):
(WebKit::WebPage::executeKeypressCommandsInternal):
(WebKit::WebPage::attributedSubstringForCharacterRangeAsync):
(WebKit::WebPage::getStringSelectionForPasteboard):
(WebKit::WebPage::getDataSelectionForPasteboard):
(WebKit::WebPage::shouldDelayWindowOrderingEvent):
(WebKit::WebPage::requestAcceptsFirstMouse):
(WebKit::WebPage::performImmediateActionHitTestAtLocation):
* Source/WebKitLegacy/mac/Misc/WebSharingServicePickerController.mm:
(-[WebSharingServicePickerController sharingService:didShareItems:]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _didConcludeEditDrag]):
(-[WebView _executeCoreCommandByName:value:]):

Canonical link: https://commits.webkit.org/275171@main
